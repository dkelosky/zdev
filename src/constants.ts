import { sep } from "path";

export interface DataSet {
    directoryBlocks?: number;
    recordFormat?: string;
    recordLength?: number;
    dataSetType?: string;
    blockSize?: number;
    size?: string;
    volumeSerial?: string;
}

// TODO(Kelosky): interface for config

export interface DataSets {
    [key: string]: DataSet;
}

// TODO(Kelosky): wrap this in a class and perhaps nested class so that we dont the error on `zdev init` command
export class Constants {

    private static _instance: Constants;
    private static readonly _QUAL = `PUBLIC`; // TODO(Kelosky): configurable
    private static readonly _CONFIG_FILE = `zdev.config.json`;
    private static readonly _CONFIG_USER_FILE = `zdev.config.user.json`;
    private static readonly _LOADLIB = "LOADLIB";
    private static readonly _HOME = "/tmp"; // TODO(Kelosky): configurable

    private _quiet: boolean;

    constructor() {
        this._quiet = false;
    }

    get user() {
        try {
            let userConfig = require(`${process.cwd()}${sep}${Constants._CONFIG_USER_FILE}`);
            return userConfig.user;
        } catch (err) {
            if (!this._quiet) {
                console.log(`üìù no ${Constants._CONFIG_USER_FILE} exists (see '${CMD_NAME} config --help' for more information)\n`);
                throw new Error("see previous message");
            }
        }
    }

    get project() {
        try {
            let projectConfig = require(`${process.cwd()}${sep}${CONFIG_FILE}`);
            return projectConfig.project;
        } catch (err) {
            if (!this._quiet) {
                console.log(`üìù no ${Constants._CONFIG_FILE} exists (see '${CMD_NAME} init --help' for more information)\n`);
                throw new Error("see previous message");
            }
        }
    }

    get hlq() {
        return this.user.toUpperCase();
    }

    get qual() {
        return Constants._QUAL;
    }

    get dsnPattern() {
        return `${this.hlq}.${this.qual}.${this.project.toUpperCase()}.`
    }

    get loadLib() {
        return this.dsnPattern + Constants._LOADLIB;
    }

    set quiet(state: boolean) {
        this._quiet = state;
    }

    get zfs() {
        return `${this.user.toUpperCase()}.${this.qual}.${this.project.toUpperCase()}.ZFS`;
    }

    get home() {
        return Constants._HOME;
    }

    get targetZfsDir() {
        return `${this.home}/${this.user.toLowerCase()}/${this.project.toLowerCase()}`;
    }

    get taretZfsDirDeploy() {
        return `${this.home}/${this.user.toLowerCase()}/${this.project.toLowerCase()}/${SOURCE_DIR}`;
    }

    get listingDir() {
        return `${LISTING_DIR}/${this.project}`;
    }

    get dataSets() {
        try {
            let projectConfig = require(`${process.cwd()}${sep}${CONFIG_FILE}`);
            return projectConfig.dataSets;
        } catch (err) {
            if (!this._quiet) {
                console.log(`üìù no ${Constants._CONFIG_FILE} exists (see '${CMD_NAME} init --help' for more information)\n`);
                throw new Error("see previous message");
            }
            return {};
        }
    }

    refresh() {
        delete require.cache[require.resolve(`${process.cwd()}${sep}${Constants._CONFIG_USER_FILE}`)];
        delete require.cache[require.resolve(`${process.cwd()}${sep}${CONFIG_FILE}`)];
    }

    static get instance() {

        if (this._instance == null) {
            this._instance = new Constants();
        }

        return this._instance;
    }

}

export const LISTING_DIR = ".listings";

export const SOURCE_DIR = `zossrc`;

export const CMD_NAME = `zdev`;
export const CACHE_NAME = `.${CMD_NAME}`;
export const WORK_COVERAGE_DIR = `${CACHE_NAME}${sep}coverage`;
export const COVERAGE_DIR = `coverage`;
export const COVERAGE_FILE = `coverage-final.json`;
export const SOURCE_CACHE_DIR_NAME = `${CACHE_NAME}${sep}${SOURCE_DIR}`;
export const COVERAGE_RESULTS_SUFFIX = `.results.txt`;
export const ADATA_SUFFIX = ".adata";
export const CACHE_SUFFIX = ".json";
export const COVERAGE_ADATA_SUFFIX = `.adata.json`;

export const HLASM_MACRO_SUFFIX = ".mac";

export const JSON_INDENT = 4;

export const BIN_SUFFIX = ".bin";
export const TXT_SUFFIX = ".txt";

export const LISTING_SUFFIX = ".lst";
export const VSCODE_FOLDER = ".vscode";
export const VSCODE_TASKS_FILE = "tasks.json";


export const CONFIG_FILE = `zdev.config.json`;
export const CONFIG_USER_FILE = `zdev.config.user.json`;

// primary command
export const ZOWE = "zowex"; // `zowe` for non-daemon

class State {

    constructor(private _debug = false) { }

    get debug() {
        return this._debug;
    }

    set debug(newState: boolean) {
        this._debug = newState;
    }
}

export const STATE = new State();