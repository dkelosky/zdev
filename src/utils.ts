import { exec } from "child_process";
import { promisify } from "util"

const exc = promisify(exec);

export async function runCmd(cmd: string, rfj = false) {
    cmd += (rfj) ? " --rfj" : ""
    const resp = await exc(cmd);
    if (resp.stderr) {
        throw new Error(resp.stderr);
    }
    return resp.stdout;
}