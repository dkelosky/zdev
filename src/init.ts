import { writeFile, exists } from "fs";
import { promisify } from "util"
import { CACHE_NAME, CMD_NAME } from "./constants";

const write = promisify(writeFile);
const exist = promisify(exists);

const FILE_NAME = "user.config.json";

export interface IOptions {
    force: boolean;
}

interface IConfig {
    user: string;
    project: string;
};

export async function init(project: string, user: string, options?: IOptions) {

    const exists = await exist(FILE_NAME);

    if (exists) {

        if (options?.force) {
            doInit(project, user);
        } else {
            console.log(`❌  Project already initialized.\n`);
            console.log(`Rerun:\n  ${CMD_NAME} init <project> --user <name> --force`);
        }

    } else {
        doInit(project, user);
    }

}

async function doInit(project: string, user: string) {
    console.log(`Initializing '${project}' with new config for '${user}'`);
    await initConfig(project, user);
    await initGitIgnore();
    console.log(`✔️  complete.`)
}

async function initConfig(project: string, user: string) {

    const config: IConfig = { user, project };

    await write(FILE_NAME, JSON.stringify(config, null, 4));
}

async function initGitIgnore() {

    // TODO(Kelosky): push in these lines if .gitignore exists
    // TODO(Kelosky): put these in a config file
    const CONTENT =
        "node_modules\n" +
        "user.config.json\n" +
        `${CACHE_NAME}\n` +
        ".listings";

    const GITIGNORE = ".gitignore";

    await write(GITIGNORE, CONTENT);

}