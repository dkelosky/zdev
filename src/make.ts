#! /bin/env node
import { runCmd } from "./utils"
import { ZOWE, TARGET_DIR } from "./constants"

(async () => {
    await make(TARGET_DIR);
})();

async function make(dir: string) {
    const makeCmd = `${ZOWE} uss issue ssh \\"cd ${dir} && make\\"`;
    const strResp = await runCmd(makeCmd);
    console.log(`... ${strResp}`);
}