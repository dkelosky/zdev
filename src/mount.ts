import { runCmd } from "./utils"
import { ZOWE} from "./constants"

export async function mountZfs(zfs: string, dir: string) {
    const mountCmd = `${ZOWE} files mount fs ${zfs} '${dir}' -m rdwr`;

    let mounted = await isMounted(zfs, dir);

    if (!mounted) {
        console.log(`Mounting ZFS "${zfs}" to "${dir}"`);
        let strResp = await runCmd(mountCmd);
        console.log(`... ${strResp}`);
    }
}

export async function isMounted(zfs: string, dir: string) {
    const listCmd = `${ZOWE} uss issue ssh 'df "${dir}"'`;

    console.log(`Checking for mount...`);
    let strResp = await runCmd(listCmd, true);
    const jsonResp = JSON.parse(strResp);

    const lines: string[] = jsonResp.stdout.split(/\r?\n/);

    let mounted = false;
    for (let i = 0; i < lines.length; i++) {
        if (lines[i].indexOf(`${zfs}`) > -1) {
            mounted = true;
            break;
        }
    }

    if (mounted) {
        console.log(`... ${dir} is mounted`);
    } else {
        console.log(`... ${dir} is not mounted`);
    }

    return mounted;
}