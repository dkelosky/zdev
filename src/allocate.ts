#! /bin/env node

import { runCmd } from "./utils"
import { ZOWE, TARGET_DIR, ZFS } from "./constants"
import { mountZfs } from "./mount"

// (async () => {
//     await createDirs(TARGET_DIR);
//     await creatZfs(ZFS);
//     await mountZfs(ZFS, TARGET_DIR);
// })();

// export _BPXK_JOBLOG=STDERR

export async function createDirs(dir: string) {
    console.log(`Creating directory "${dir}"...`);
    const mkdirCmd = `${ZOWE} uss issue ssh 'mkdir -p "${dir}"'`;
    const strResp = await runCmd(mkdirCmd);
    console.log(`${strResp}`);
}

export async function creatZfs(zfs: string) {
    const listCmd = `${ZOWE} files list ds "${ZFS}"`;
    const createCmd = `${ZOWE} files create zfs "${ZFS}"`;

    const volumes = `--volumes `

    console.log(`Checking for ZFS "${zfs}"...`);
    let strResp = await runCmd(listCmd, true);
    const jsonResp = JSON.parse(strResp);

    if (jsonResp.data.apiResponse.returnedRows === 0) {
        strResp = await runCmd(createCmd);
        console.log(`... ${strResp}`);
    } else {
        console.log(`... data set already exists`);
    }
}
