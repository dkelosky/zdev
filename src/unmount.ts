import { runCmd } from "./utils";
import { ZOWE, ZFS, TARGET_DIR } from "./constants";
import { isMounted } from "./mount";

(async () => {
    await unmount(ZFS);
})();

async function unmount(zfs: string) {

    let mounted = await isMounted(ZFS, TARGET_DIR);
    if (mounted) {
        console.log(`Unmounting "${zfs}"...`);
        const unmountCmd = `${ZOWE} files unmount fs "${zfs}"`;
        try {
            const strResp = await runCmd(unmountCmd);
            console.log(`${strResp}`);
        } catch (err) {
            console.log(err)
        }
    }

}
