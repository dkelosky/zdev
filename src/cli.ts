#! /bin/env node

// import * as cmd from "commander";
// const cli = new cmd.Command();
import { createDirs, creatZfs } from "./actions/allocate-zfs";
import { init } from "./init";
import { mountZfs } from "./actions/mount";
import { unmount } from "./actions/unmount";
import { uploadAll, uploadChanged, uploadFiles } from "./actions/zfs-upload"
import { version, command, parse, help, Command, description, option } from "commander";
import { TARGET_ZFS_DIR, ZFS, CMD_NAME, TARGET_ZFS_DIR_DEPLOY, SOURCE_DIR } from "./constants"
import { make } from "./actions/make";
import { getDirs, getListings, runCmd } from "./utils";
import { run } from "./actions/run";

// BPXBATCH - ZOA
// https://www.ibm.com/docs/en/zos/2.3.0?topic=functions-dynalloc-allocate-data-set
// https://www.ibm.com/docs/en/zos/2.2.0?topic=output-example-calling-bpxwdyn-from-c
// https://www.ibm.com/docs/en/zos/2.1.0?topic=descriptions-extattr-set-reset-display-extended-attributes-files
// https://www.ibm.com/docs/en/zos/2.1.0?topic=descriptions-exec-bpx1exc-bpx4exc-run-program

// TODO(Kelosky): template files
// TODO(Kelosky): help with asmchdrs
// TODO(Kelosky): build command that can once per day allocate data sets and cache that info
// TODO(Kelosky): make should trigger allocate conditionally and upload if out of sync
// TODO(Kelosky): create tests on directory so things can be moved without breaking!!
// TODO(Kelosky): sync command to delete old files not needed
// TODO(Kelosky): add lib & init lib command so helper code can be shared with projects
// TODO(Kelosky): fallback for batch / JCL to build

//DSECT4 EXEC PGM=CCNEDSCT,
//         PARM='&DPARM,SECT(ALL)',
//         MEMLIMIT=256M
//STEPLIB  DD  DISP=SHR,DSN=CEE.SCEERUN2
//         DD  DISP=SHR,DSN=CBC.SCCNCMP
//         DD  DISP=SHR,DSN=CEE.SCEERUN
//SYSADATA DD  DISP=SHR,DSN=KELDA16.PUBLIC.MTL.ADATA(IHAASCB)
//EDCDSECT DD  DISP=SHR,DSN=KELDA16.PUBLIC.MTL.CHDR(IHAASCB)
//SYSPRINT DD SYSOUT=*
//SYSOUT   DD SYSOUT=*

version(`0.0.1`)
description(`Example:\n` +
    `  ${CMD_NAME} init <name> -u kelda16\n` +
    `  ${CMD_NAME} allocate\n`
);

command(`init <project>`)
    .requiredOption(`-u, --user <name>`)
    .option(`-f, --force`)
    .description(`init a project`)
    .action(async (project: string, options: any, cmd: Command) => {
        await init(project, options.user, { force: options.force });
    });

// TODO(Kelosky): for these test for user = IBMUSER
// TODO(Kelosky): test for z/osmf profile
// TODO(Kelosky): on fresh create zfs, clear cache if it exists

command(`allocate`)
    .description(`allocate zfs`)
    .action(async () => {
        await createDirs(TARGET_ZFS_DIR);
        await creatZfs(ZFS);
        await mountZfs(ZFS, TARGET_ZFS_DIR);

        const list = await getDirs(SOURCE_DIR);

        for (let i = 0; i < list.length; i++) {
            await createDirs(`${TARGET_ZFS_DIR}/${list[i]}`);
        }

    });

command(`x`)
    .description(`zdev testing`)
    .action(async () => {

        const listing =
        `xlc++ -W "c,lp64,langlvl(extended),xplink,exportall" -qsource  -g  -c -qlist=lib/run.cpp.lst -o lib/run.o lib/run.cpp\n` +
        `"lib/runexec.h", line 1.0: CCN5809 (W) The source file is empty.\n` +
        `"lib/run.cpp", line 6.5: CCN8145 (W) "main" cannot be exported. The directive is ignored.\n` +
        `xlc -S -W "c,metal, langlvl(extended), sscom, nolongname, inline, genasm, inlrpt, csect, nose, list, optimize(2), list, showinc, showmacro, source, aggregate" -qlist=lib/runexec.mtl.lst -I/usr/include/metal  -Ilib/ -o lib/runexec.s lib/runexec.c\n` +
        `WARNING CCN3229 lib/runexec.h:1     File is empty.\n` +
        `as  -a=lib/runexec.asm.lst -ISYS1.MACLIB  -ICBC.SCCNSAM -o lib/runexec.o lib/runexec.s\n` +
        ` Assembler Done No Statements Flagged\n` +
        `xlc++ -W "l,lp64,dll,dynam=dll,xplink,map,list"  -g  -qsource -o lib/run lib/run.o lib/runexec.o > lib/run.bind.lst\n` +
        `IGD01008I ALLOCATION SET TO SCTEMPD M0610\n` +
        `IGD01010I ALLOCATION SET TO SGTEMPD STORAGE GROUP\n` +
        `IGD01008I ALLOCATION SET TO SCTEMPD M0610\n` +
        `IGD01010I ALLOCATION SET TO SGTEMPD STORAGE GROUP\n`;

        const list = await getListings(listing);
        console.log(list);
    });

command(`run <target>`)
    .description(`run a file`)
    .action(async (target: string) => {
        await run(target);
    });

command(`make [target]`)
    .description(`run make`)
    .action(async (target: string) => {
        await make(target || "");
    });

command(`mount`)
    .description(`mount zfs`)
    .action(async () => {
        await mountZfs(ZFS, TARGET_ZFS_DIR);
    });

command(`unmount`)
    .description(`unmount zfs`)
    .action(async () => {
        await unmount(ZFS);
    });

command(`upload [files...]`)
    .description(`upload zfs files\n  zdev upload\n  zdev upload main.s lib/ecbwait.h`)
    .option(`-f, --force`)
    .action(async (files: string[], options: any, cmd: Command) => {
        if (options.force) {
            await uploadAll();
        } else if (files.length > 0) {
            await uploadFiles(files);
        } else {
            await uploadChanged();
        }
    });

const cmd = parse(process.argv);

if (cmd.args.length === 0) {
    help();
}



