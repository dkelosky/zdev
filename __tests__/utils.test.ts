import Mock = jest.Mock;
jest.mock("fs");

import { getDirs, getListings } from "../src/utils";
import { stat, readdir, Stats } from "fs";

describe("utils tests", () => {

    describe("getListings tests", () => {
        it("should parse metal, asm, and binder listings", async () => {

            const text =
                `xlc -S -W "c,metal, langlvl(extended), sscom, nolongname, inline, genasm, inlrpt, csect, nose, lp64, list, warn64, optimize(2), list, showinc, showmacro, source, aggregate" -qlist=mtlmain.mtl.lst -I/usr/include/metal -o mtlmain.s mtlmain.c\n` +
                `WARNING CCN3296 ./wto.h:15    #include file "ihaecb.h" not found.\n` +
                `as  -a=mtlmain.asm.lst -o mtlmain.o mtlmain.s\n` +
                ` Assembler Done No Statements Flagged\n` +
                `ld -bRMODE=ANY -V -eMAIN -o mtlmain mtlmain.o > mtlmain.bind.lst\n` +
                ` IEW2278I B352 INVOCATION PARAMETERS -\n` +
                `          TERM=YES,PRINT=NO,MSGLEVEL=4,STORENX=NEVER,RMODE=ANY,LIST=NOIMP,XREF=\n` +
                `          YES,MAP=YES,PRINT=YES,MSGLEVEL=0\n` +
                ` IEW2008I 0F03 PROCESSING COMPLETED.  RETURN CODE =  0.\n`;

            const listings = await getListings(text);
            expect(listings).toMatchSnapshot();
        });
    });

    // describe("createCacheDirs tests", () => {

    //     const fn = (stat as any) as Mock<ReturnType<typeof stat>, Parameters<typeof stat>>;
    //     fn.mockImplementation((path) => {
    //         return {
    //             is
    //         }
    //     });

    // });

    describe("getDirs tests", () => {

        it("should get a list of dirs from a starting dir location recursively", async () => {

            let depth = 3; // number of /dir in path

            // const rd = (readdir as any) as Mock<ReturnType<typeof readdir>, Parameters<typeof readdir>>;
            const rd = (readdir as any) as Mock<ReturnType<typeof readdir>, any>;
            rd.mockImplementation((path, cb) => {
                cb(null, ["file1", "file2", "dir"]);
                depth--;
            });

            const st = (stat as any) as Mock<ReturnType<typeof stat>, any>;
            // const st = (stat as any) as Mock<ReturnType<typeof stat>, Parameters<typeof stat>>;
            st.mockImplementation((path: string, cb) => {

                const sts: Stats = {
                    isDirectory: () => {
                        if (path.indexOf("dir") > -1 && depth > -1 && path.indexOf("file") === -1) {
                            return true;
                        } else {
                            return false;
                        }
                    }
                } as any;

                cb(null, sts);

            });

            const dirs = await getDirs(`some/location`);
            expect(dirs).toMatchSnapshot();
        });


    });

});