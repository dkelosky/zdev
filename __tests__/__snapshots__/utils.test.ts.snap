// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils tests getDirs tests should get a list of dirs from a starting dir location recursively 1`] = `
Array [
  "some/location",
  "some/location/dir",
  "some/location/dir/dir",
  "some/location/dir/dir/dir",
]
`;

exports[`utils tests getListings tests should parse metal, asm, and binder listings 1`] = `
Array [
  "mtlmain.mtl.lst",
  "mtlmain.asm.lst",
  "mtlmain.bind.lst",
]
`;
